notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bdf9ad82-4098-4fab-bc7d-3b91068ee73e",
+   "metadata": {},
+   "source": [
+    "MSQ has an inherent disadvantage over Scatter/Gather when it comes to response time for many queries. It must first initialize the `query_controller` and `query_worker` tasks and only then begin to process the query.  You must also make sure that you have enough worker slots available for them. If any of the tasks are waiting to be assigned, the whole query will wait before it starts executing stages. It will fail after a timeout if worker slots do not become available.\n",
+    "MSQ reads the segment files directly from Deep Storage which is also typically slower than the local storage that Historicals use and there is no caching of segment files. It has the advantage that it can access data that is not cached in the historicals but still exists in Deep Storage. So it is best for longer running queries on broader or older timeframes.\n",
+    "\n",
+    "As queries are running you can see the `query_controller` and `query_worker` tasks being spawned for each query in the [tasks view of the Druid Console](http://localhost:8888/unified-console.html#tasks).\n"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "ab3a31ef-45aa-41e3-9708-4e2ec0488600",
+   "metadata": {},
+   "source": [
+    "### Sort Merge joins in MSQ\n",
+    "SortMerge joins execute by hash distributing both sides of the join among all the `query_workers` using the join columns to calculate the hash. The end result is that the rows from each of the tables that correspond to a given join column value will end up on the same worker:\n",
+    "![](assets/msq-sortmerge.png)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "id": "554204b6-df63-4c14-96e8-4341d1b7fa8f",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:12015.551500000001 ms   min:11389.986 ms  max:13015.264000000001 ms\n",
+      "2023-10-26 01:06:05.324785 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 6, 5, 324785)"
+      ]
+     },
+     "execution_count": 26,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "#MSQ async query using sortMerge\n",
+    "print(measure_async_query(sql,10,joinAlgorithm='sortMerge'))\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "88a0f948-3a9e-405b-a305-95fed27db6ca",
+   "metadata": {},
+   "source": [
+    "In this case, sortMerge joins did not help which makes sense. A broadcast of 4000 `users` rows to the workers is much faster than having to redistribute both `clicks` and `users`. But if the two tables are large, a broadcast of one of them will not be possible because it would not fit in the Heap of the `query_workers`.\n",
```

Comment:
Can you demonstrate this with an example?

Reply:
I do show it by joining click with click in the next section.